### PR TITLE
Pagination in urls page

### PR DESCRIPTION
**Pull request by jatinkatyal13**

### CODE REVIEW

File: src/controllers/urls.ts [hunk 1]
```diff
@@ -72,11 +84,22 @@ export const findUrlByCodeInt = async (codeInt: number) =>
...
+) => {
+  const options = {
     where: {
       ownerId: user.id,
     },
-  })
-  return urls!
+    ...limit
+  }
+  const [urls, urlCount] = await Promise.all([URLs.findAll(options), URLs.count()])
```
Comment (championswimmer) [inline src/controllers/urls.ts]: sequelize has a `findAndCountAll` method to do this in a single transaction

File: src/controllers/urls.ts [hunk 2]
```diff
@@ -10,6 +10,18 @@ export interface URLOptions {
   shortCode?: string
 }
 
+export interface LimitingOptions {
```
Comment (championswimmer) [inline src/controllers/urls.ts]: should ideally call this `PageOptions`

File: src/controllers/urls.ts [hunk 3]
```diff
@@ -72,11 +84,22 @@ export const findUrlByCodeInt = async (codeInt: number) =>
     },
   })
 
-export const getAllUrlsForUser = async (user: UserAttributes) => {
-  const urls = await URLs.findAll({
+export const getAllUrlsForUser = async (
+  user: UserAttributes, 
+  limit: LimitingOptions
```
Comment (championswimmer) [inline src/controllers/urls.ts]: this param should be called `page` and not `limit` as offset and **limit** are two parts of page options

File: views/pages/urls/index.hbs [hunk 4]
```diff
@@ -1,5 +1,19 @@
 {{!< center-container }}
 
+<nav aria-label="Page navigation example">
+    <ul class="pagination justify-content-end">
+        <li class="page-item {{#unless pagination.hasPrev}}disabled{{/unless}}">
+            <a class="page-link" href="?page={{ dec pagination.page }}" tabindex="-1">Previous</a>
+        </li>
+        {{#range pagination.pageCount}}
```
Comment (championswimmer) [inline views/pages/urls/index.hbs]: do not add helpers unnecessarily too much. range is a bit pythonish way of doing things. we can simple achieve this using the foreach helpers that are inbuilt in hbs

Comment (jatinkatyal13) [reply views/pages/urls/index.hbs]: Actually pagecount is a number object here and foreach helper in hbs only iterates over iterables! That is why I needed to add this !

**Comment by jatinkatyal13** [comment]
Required changes made 🚀

File: views/pages/urls/index.hbs (hunk 4 above)
Comment (championswimmer) [reply views/pages/urls/index.hbs]: Massage the data into an iterable ;)

Comment (jatinkatyal13) [reply views/pages/urls/index.hbs]: Done 🚀

File: src/routes/pages/urls.ts [hunk 5]
```diff
@@ -5,16 +5,33 @@ import {
...
+
 route.get('/', async (req, res) => {
-  const urls = await getAllUrlsForUser(req.user)
-  return res.render('pages/urls/index', { urls })
+  const page: PageOptions = {
+    offset: req.query.limit * (req.query.page - 1),
+    limit: req.query.limit,
+  }
+  const { urls, pagination } = await getAllUrlsForUser(req.user, page)
+  const pageList = [...Array(pagination.pageCount).keys()]
```
Comment (championswimmer) [inline src/routes/pages/urls.ts]: nitpick, but the best method to generate this array is 

```js
Array.from({length: pagination.pageCount}, (v, k) => k+1);
```
https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/from

Comment (jatinkatyal13) [reply src/routes/pages/urls.ts]: @championswimmer Donzo 😄 🚀

File: src/controllers/urls.ts [hunk 6]
```diff
@@ -73,11 +85,22 @@ export const findUrlByCodeInt = async (codeInt: number) =>
...
+) => {
+  const options = {
     where: {
       ownerId: user.id,
     },
-  })
-  return urls!
+    ...page,
+  }
+  const { rows, count } = await URLs.findAndCountAll(options)
```
Comment (championswimmer) [inline src/controllers/urls.ts]:
```ts
let {offset = 0, limit = 50} = page
```

File: src/controllers/urls.ts [hunk 7]
```diff
@@ -11,6 +11,18 @@ export interface URLOptions {
   shortCode?: string
 }
 
+export interface PageOptions {
+  offset: number
```
Comment (championswimmer) [inline src/controllers/urls.ts]: offset?
limit?

File: src/controllers/urls.ts [hunk 8]
```diff
@@ -73,11 +85,22 @@ export const findUrlByCodeInt = async (codeInt: number) =>
     },
   })
 
-export const getAllUrlsForUser = async (user: UserAttributes) => {
-  const urls = await URLs.findAll({
+export const getAllUrlsForUser = async (
+  user: UserAttributes,
+  page: PageOptions,
```
Comment (championswimmer) [inline src/controllers/urls.ts]: page: PageOptions = {}

File: src/controllers/urls.ts [hunk 9]
```diff
@@ -73,11 +85,22 @@ export const findUrlByCodeInt = async (codeInt: number) =>
...
       ownerId: user.id,
     },
-  })
-  return urls!
+    ...page,
+  }
+  const { rows, count } = await URLs.findAndCountAll(options)
+  const pagination: PaginationOptions = {
+    page: Math.floor(page.offset / 20) + 1,
+    pageCount: Math.ceil(count / 20),
```
Comment (championswimmer) [inline src/controllers/urls.ts]: use the value in limit instead of hardcode to 20

**Pull request merged**